### PR TITLE
Report Installed Packages

### PR DESCRIPTION
**Pull request by kevensen**
This feature reports the installed packages, either rpm or dpkg, in the image under inspection.  The report is presented in JSON at api/v1/packages.

### CODE REVIEW

**Comment by simon3z** [comment]
@enoodle @moolitayer @pweil- can you review?

**Comment by moolitayer** [comment]
We should run this scan type only when it is requested by an options.
it should not run in the ManageIQ flow for example.

File: pkg/content/package.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	"path"
+	"strings"
+)
+
+type defaultPackageInspector struct {
+	packageInfo         chrootContentFunc
+	buildPackageCommand chrootCommandFunc
+	cmd                 *exec.Cmd
+
+	Packages []packages
```
Comment (pweil-) [inline pkg/content/package.go]: public field on a private struct

File: pkg/content/package.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	"bytes"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+)
+
+type defaultPackageInspector struct {
```
Comment (pweil-) [inline pkg/content/package.go]: godoc for structs and fields please

File: pkg/content/package.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+type packages struct {
+	Package      string
+	Version      string
+	Architecture string
+	Description  string
+}
+
+// NewDefaultPackageInspector returns a new Version Inspector
+func NewDefaultPackageInspector() Inspector {
+	inspector := &defaultPackageInspector{}
```
Comment (pweil-) [inline pkg/content/package.go]: return &defaultPackageInspector{}

File: pkg/content/package.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+	return inspector
+}
+
+//buildPathChroot Builds the new PATH environment variable for executing
+//commands
+func buildPathChroot(mountPath string) string {
+
```
Comment (pweil-) [inline pkg/content/package.go]: dead space

File: pkg/content/package.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	cmd := exec.Command("rpm", "-qa", "--dbpath",
+		path.Join(mountPath, "/var/lib/rpm"), "--qf",
+		"%{NAME},%{VERSION},%{ARCH},%{DESCRIPTION}---")
+	cmd.Env = []string{buildPathChroot(mountPath)}
+	s.cmd = cmd
+}
+
+//getPackageInformation executes the appropriate command and parses the output
+func (s *defaultPackageInspector) getPackageInformation() error {
+
```
Comment (pweil-) [inline pkg/content/package.go]: dead space

File: pkg/content/package.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	out, err := s.cmd.CombinedOutput()
+
+	if err != nil {
+		return fmt.Errorf("Unable to read package content")
+	}
+	pkgdata := strings.Replace(string(out), "\n", "", -1)
+	pkglist := strings.Split(pkgdata, "---")
```
Comment (pweil-) [inline pkg/content/package.go]: what happens when someone decides to put `---` in a description or is this a value that is not allowed in descriptions?

File: pkg/content/package.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+				Version: parts[1], Architecture: parts[2], Description: pkgDescription})
+		}
+	}
+
+	return nil
+}
+
+//Inspect is the primary execution entrypoint
+func (s *defaultPackageInspector) Inspect(mountPath string) error {
+	_, err := os.Stat(path.Join(mountPath, "/var/lib/rpm"))
```
Comment (pweil-) [inline pkg/content/package.go]: should we rely on the hard coded binary or resolve this via the path variable for the image?

File: pkg/content/package.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+	out, err := s.cmd.CombinedOutput()
+
+	if err != nil {
+		return fmt.Errorf("Unable to read package content")
+	}
+	pkgdata := strings.Replace(string(out), "\n", "", -1)
+	pkglist := strings.Split(pkgdata, "---")
+
+	for _, pkg := range pkglist {
```
Comment (pweil-) [inline pkg/content/package.go]: separate parsing into a new method that takes a string so you can pass in data to be able to unit test your parsing code

File: pkg/inspector/image-inspector.go [hunk 9]
```diff
@@ -123,6 +125,13 @@ func (i *defaultImageInspector) Inspect() error {
 		}
 	}
 
+	packageInspector := content.NewDefaultPackageInspector()
```
Comment (pweil-) [inline pkg/inspector/image-inspector.go]: new scan type.  Also means we probably need to factor the serve portion of the code a little better to be able to serve based on requested types if it is not already doing that.

**Comment by kevensen** [comment]
@moolitayer - I think you are right.  There may be a scenario where one would want to execute an out-of-band scan including both and OpenSCAP report and Package report.  The current "scan-type" option only allows for one type at a time.

Perhaps we should turn each scan type into a boolean?  --openscap --packagreport?  If you don't mind, I think I'll add an option for --packagereport.  That way, it won't run by default (e.g. during a ManageIQ SmartState).  But, I am open to recommendations.

File: pkg/content/package.go (hunk 6 above)
Comment (kevensen) [inline pkg/content/package.go]: That was a gamble (a bad gamble) that no one would put "---" in their description.  I'll have to think of a smarter way to parse this.

**Comment by kevensen** [comment]
@pweil- I went back to the drawing board and reworked this feature (as well as added some unit tests).  This should behave more predictably.

**Comment by kevensen** [comment]
This is complaining about not having a Rakefile.  Do you guys have a travis.yml file you use for CI?

**Comment by enoodle** [comment]
@kevensen We want to add it on #27

**Comment by kevensen** [comment]
@enoodle - Just so I understand, you intend to merge #27 first?  Would you like me to close this pull request and open a new one after #27 is merged?  

If so, that's just fine.   Just let me know what you'd like me to do.

**Comment by simon3z** [comment]
Before we move on with this one (more work, etc.) I should review the idea behind this.
There were some consideration exactly on why not going in this direction.

**Comment by kevensen** [comment]
@simon3z - I know that CloudForms performs the SmartState analysis which provides the information this feature presents.  In the case the image-inspector (a very useful tool btw) is used outside of CloudForms, I believe that being able to assess what is in an image is invaluable.  In any case, I am happy to discuss the design considerations.

**Comment by simon3z** [comment]
Sorry I feel this is out of scope at the moment. Closing this for now. Feel free to reopen if needed.

**Pull request closed without merging**